packages/gatsby/src/schema/infer/infer.js

```diff
@@ -0,0 +1,302 @@
+          })
+        }
+      }
+      let field = typeComposer.getField(key)
+      if (getNamedType(fieldType).name === fieldConfig.type) {
+        if (!field.type) {
+          field = {
+            type: field,
+          }
+        }
+        if (field.type === fieldConfig.type)
+          if (!field.args && fieldConfig.args) {
+            field.args = fieldConfig.args
+          }
+        if (!field.resolve && fieldConfig.resolve) {
+          field.resolve = fieldConfig.resolve
+        }
+        typeComposer.setField(key, field)
+      }
+    } else {
+      fields[key] = fieldConfig
+    }
+  })
+
+  Object.keys(fields).forEach(fieldName => {
+    typeComposer.setField(fieldName, fields[fieldName])
+  })
+  return typeComposer
+}
+
+// XXX(freiksenet): removing this as it's a breaking change
+// Deeper nested levels should be inferred as JSON.
+// const MAX_DEPTH = 5
+
+const hasMapping = selector => {
+  const { mapping } = store.getState().config
+  return mapping && Object.keys(mapping).includes(selector)
+}
+
+const getFieldConfigFromMapping = selector => {
+  const { mapping } = store.getState().config
+  const [type, ...path] = mapping[selector].split(`.`)
+  return { type, resolve: link({ by: path.join(`.`) || `id` }) }
+}
+
+// probably should be in example value
+const getFieldConfigFromFieldNameConvention = (
+  schemaComposer,
+  nodeStore,
+  value,
+  key
+) => {
+  const path = key.split(`___NODE___`)[1]
+  // Allow linking by nested fields, e.g. `author___NODE___contact___email`
+  const foreignKey = path && path.replace(/___/g, `.`)
+
+  const getNodeBy = value =>
+    foreignKey
+      ? nodeStore.getNodes().find(node => _.get(node, foreignKey) === value)
+      : nodeStore.getNode(value)
+
+  const linkedNodes = Array.isArray(value)
+    ? value.map(getNodeBy)
+    : [getNodeBy(value)]
+
+  const linkedTypes = _.uniq(
+    linkedNodes.filter(Boolean).map(node => node.internal.type)
+  )
+
+  invariant(
+    linkedTypes.length,
+    `Could not infer a GraphQL type for the field "${key}".`
+  )
+
+  let type
+  // If the field value is an array that links to more than one type,
+  // create a GraphQLUnionType. Note that we don't support the case where
+  // scalar fields link to different types. Similarly, an array of objects
+  // with foreign-key fields will produce union types if those foreign-key
+  // fields are arrays, but not if they are scalars. See the tests for an example.
+  // FIXME: The naming of union types is a breaking change. In current master,
+  // the type name includes the key, which is (i) potentially not unique, and
+  // (ii) hinders reusing types.
+  if (linkedTypes.length > 1) {
+    const typeName = linkedTypes.sort().join(``) + `Union`
+    schemaComposer.getOrCreateUTC(typeName, utc => {
+      const types = linkedTypes.map(typeName =>
+        schemaComposer.getOrCreateTC(typeName)
+      )
+      utc.setTypes(types)
+      types.forEach(type => {
+        utc.addTypeResolver(
```

Comment:
In graphql-compose 5.8 we now have `utc.setResolveType(node => node.internal.type)`

Reply:
Yeah, same for interfaces. I've actually asked maintainer to add that :D I'll fix that.